src/box/tuple.h

```diff
@@ -334,19 +334,26 @@ struct PACKED tuple
 	/** Format identifier. */
 	uint16_t format_id;
 	/**
-	 * Length of the MessagePack data in raw part of the tuple.
-	 * Must be set in tuple_init_default and accessed by tuple_bsize etc.
+	 * A pair of fields for storing data offset and data bsize:
+	 * Data offset - to the MessagePack from the begin of the tuple.
+	 * Data bsize - length of the MessagePack data in raw part of the tuple.
+	 * These two values can be stored in bulky and compact modes.
+	 * In bulky mode offset is simple stored in data_offset_bsize_raw
+	 * member while bsize is stored in bsize_bulky.
+	 * In compact mode bsize_bulky is not used and both values are combined
+	 * in data_offset_bsize_raw: offset in high byte, bsize in low byte,
+	 * and the highest bit is set to 1 as a flag of compact mode.
+	 * They're for internal use only, must be set in tuple_init_default and
+	 * accessed by tuple_data_offset, tuple_bsize etc.
 	 */
-	uint32_t bsize;
-	/**
-	 * Offset to the MessagePack from the begin of the tuple.
-	 * Must be set in tuple_init_default and accessed by tuple_data_offset.
-	 */
-	uint16_t data_offset;
+	uint16_t data_offset_bsize_raw;
+	uint32_t bsize_bulky;
```

Comment:
Can we use a union here to improve readability? Something like

```C
union {
    struct {
        uint8_t data_offset;
        uint8_t bsize;
    } compact;
    struct {
        uint16_t data_offset;
        uint16_t bsize;
    } bulky;
};
```

Reply:
No. 1) we have store flag here too (see my reply above). 2) That makes impossible to make branchless  data_offset getters.